monix-reactive/shared/src/main/scala/monix/reactive/internal/builders/Interleave2Observable.scala

```diff
@@ -0,0 +1,215 @@
+      lastAck = lastAck match {
+        case Continue ⇒ rawOnNext(a)
+        case Stop ⇒ if (select == 0) rawOnNext(a) else Stop
+        case async ⇒
+          async.flatMap {
+            case Continue ⇒ self.synchronized(rawOnNext(a))
+            case Stop     ⇒ self.synchronized(if (select == 0) rawOnNext(a) else Stop)
+          }
+      }
+
+      continueA1.tryCompleteWith(lastAck)
+      continueA1 = Promise[Ack]()
+      lastAck
+    }
+
+    def signalOnNextA2(a: A): Future[Ack] = {
+      lastAck = lastAck match {
+        case Continue ⇒ rawOnNext(a)
+        case Stop ⇒ if (select == 0) rawOnNext(a) else Stop
+        case async ⇒
+          async.flatMap {
+            case Continue ⇒ self.synchronized(rawOnNext(a))
+            case Stop     ⇒ self.synchronized(if (select == 0) rawOnNext(a) else Stop)
+          }
+      }
+
+      continueA2.tryCompleteWith(lastAck)
+      continueA2 = Promise[Ack]()
+      lastAck
+    }
+
+    def signalOnError(ex: Throwable): Unit = self.synchronized {
+      if (!isDone) {
+        isDone = true
+        out.onError(ex)
+        lastAck = Stop
+        continueA1.tryCompleteWith(Stop)
+        continueA2.tryCompleteWith(Stop)
+      }
+    }
+
+    def signalOnComplete(ack: Future[Ack]): Unit = {
+      @inline def rawOnComplete(): Unit =
+        if (!isDone) {
+          isDone = true
+          out.onComplete()
+        }
+
+      self.synchronized  {
+        val shouldComplete = !isDone && (
+          {
+            completedCount += 1
+            completedCount == 2
+          })
+
+        if (shouldComplete) {
+          ack match {
+            case Continue ⇒ rawOnComplete()
+            case Stop     ⇒ () // do nothing
+            case async    ⇒
+              async.onComplete {
+                case Success(Continue) ⇒
+                  self.synchronized(rawOnComplete())
+                case _                 ⇒
+                  () // do nothing
+              }
+          }
+
+          lastAck = Stop
+        }
+      }
+    }
+
+    val composite = CompositeCancelable()
+
+    composite += obsA1.unsafeSubscribeFn(new Subscriber[A] {
+      implicit val scheduler = out.scheduler
+
+      def onNext(elem: A): Future[Ack] = self.synchronized {
+        if (isDone) Stop else {
+          hasElemA1 = true
+          if (select == 0) select = 1
+
+          lastAck1 = if (select == 1)
+            signalOnNextA1(elem)
+          else
+          // here essentially means waiting until obsA2 is done with its work
+          // when that happens, before passing the `Continue` signal to the upstream observable,
+          // immediately signal `onNext` of the out Observable to emit the elem since it is already available
+            continueA2.future.flatMap {
+              case Continue ⇒
+                self.synchronized { signalOnNextA1(elem) }
+              case Stop     ⇒
+                self.synchronized {
+                  if (select >= 0)
```

Comment:
You should be careful here. I don't think there should ever be a case where `select < 0` in this `if` and if it is, then it's a bug, because you just received an `onNext` after (I suppose) an `onComplete`. And if you have a bug in this code, you're not going to see it. IMHO, I would get rid of the `if (select >= 0)` and make sure that the opposite never happens by watching the code really hard :-))